## 1. Dataset Understanding

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("data/orders.csv")

In [3]:
df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,1.0,29.99,8.72,38.71,1.0,credit_card,4.0
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,1.0,118.70,22.76,141.46,1.0,boleto,4.0
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,1.0,159.90,19.22,179.12,3.0,credit_card,5.0
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00,1.0,45.00,27.20,72.20,1.0,credit_card,5.0
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00,1.0,19.90,8.72,28.62,1.0,credit_card,5.0


In [4]:
df.tail()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
99436,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00,1.0,72.00,13.08,85.08,3.0,credit_card,5.0
99437,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00,1.0,174.90,20.10,195.00,3.0,credit_card,4.0
99438,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00,1.0,205.99,65.02,271.01,5.0,credit_card,5.0
99439,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00,2.0,359.98,81.18,441.16,4.0,credit_card,2.0
99440,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-03-08 20:57:30,2018-03-09 11:20:28,2018-03-09 22:11:59,2018-03-16 13:08:30,2018-04-03 00:00:00,1.0,68.50,18.36,86.86,1.0,debit_card,5.0


In [5]:
df.shape

(99441, 15)

In [6]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 99441
Number of columns: 15


### Observation

The dataset contains 99441 records and 15 variables.
Each record represents an e-commerce order-level observation.

In [7]:
df.columns.tolist()

['order_id',
 'customer_id',
 'order_status',
 'order_purchase_timestamp',
 'order_approved_at',
 'order_delivered_carrier_date',
 'order_delivered_customer_date',
 'order_estimated_delivery_date',
 'n_items',
 'price',
 'freight',
 'payment',
 'installments',
 'payment_type',
 'review_score']

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   order_id                       99441 non-null  object 
 1   customer_id                    99441 non-null  object 
 2   order_status                   99441 non-null  object 
 3   order_purchase_timestamp       99441 non-null  object 
 4   order_approved_at              99281 non-null  object 
 5   order_delivered_carrier_date   97658 non-null  object 
 6   order_delivered_customer_date  96476 non-null  object 
 7   order_estimated_delivery_date  99441 non-null  object 
 8   n_items                        98666 non-null  float64
 9   price                          98666 non-null  float64
 10  freight                        98666 non-null  float64
 11  payment                        99440 non-null  float64
 12  installments                   99440 non-null 

## 2. Column Classification

In [9]:
df.select_dtypes(include=np.number).columns.tolist()

['n_items', 'price', 'freight', 'payment', 'installments', 'review_score']

In [10]:
df.select_dtypes(include="object").columns.tolist()

['order_id',
 'customer_id',
 'order_status',
 'order_purchase_timestamp',
 'order_approved_at',
 'order_delivered_carrier_date',
 'order_delivered_customer_date',
 'order_estimated_delivery_date',
 'payment_type']

In [11]:
df.describe()

,n_items,price,freight,payment,installments,review_score
count,98666.000000,98666.000000,98666.000000,99440.000000,99440.000000,98673.000000
mean,1.141731,137.754076,22.871095,160.990267,2.930792,4.086793
std,0.538452,210.645145,21.825135,221.951257,2.715942,1.346274
min,1.000000,0.850000,0.000000,0.000000,0.000000,1.000000
25%,1.000000,45.900000,13.850000,62.010000,1.000000,4.000000
50%,1.000000,86.900000,17.180000,105.290000,2.000000,5.000000
75%,1.000000,149.900000,24.060000,176.970000,4.000000,5.000000
max,21.000000,13440.000000,1794.960000,13664.080000,24.000000,5.000000


In [12]:
df.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,n_items,price,freight,payment,installments,payment_type,review_score
count,99441,99441,99441,99441,99281,97658,96476,99441,98666.000000,98666.000000,98666.000000,99440.000000,99440.000000,99440,98673.000000
unique,99441,99441,8,98875,90733,81018,95664,459,NaN,NaN,NaN,NaN,NaN,5,NaN
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-04-11 10:48:14,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-08 23:38:46,2017-12-20 00:00:00,NaN,NaN,NaN,NaN,NaN,credit_card,NaN
freq,1,1,96478,3,9,47,3,522,NaN,NaN,NaN,NaN,NaN,75387,NaN
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.141731,137.754076,22.871095,160.990267,2.930792,NaN,4.086793
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.538452,210.645145,21.825135,221.951257,2.715942,NaN,1.346274
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,0.850000,0.000000,0.000000,0.000000,NaN,1.000000
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,45.900000,13.850000,62.010000,1.000000,NaN,4.000000
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,86.900000,17.180000,105.290000,2.000000,NaN,5.000000
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.000000,149.900000,24.060000,176.970000,4.000000,NaN,5.000000


In [13]:
missing = df.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

order_delivered_customer_date    2965
order_delivered_carrier_date     1783
n_items                           775
price                             775
freight                           775
review_score                      768
order_approved_at                 160
payment                             1
installments                        1
payment_type                        1
dtype: int64

In [14]:
missing_percentage = (df.isnull().sum() / len(df) * 100)

missing_percentage[missing_percentage > 0].sort_values(ascending=False)

order_delivered_customer_date    2.981668
order_delivered_carrier_date     1.793023
n_items                          0.779357
price                            0.779357
freight                          0.779357
review_score                     0.772317
order_approved_at                0.160899
payment                          0.001006
installments                     0.001006
payment_type                     0.001006
dtype: float64

In [15]:
missing_report = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Missing Percentage": df.isnull().mean() * 100
})

missing_report = missing_report[
    missing_report["Missing Values"] > 0
].sort_values("Missing Values", ascending=False)

missing_report

,Missing Values,Missing Percentage
order_delivered_customer_date,2965,2.981668
order_delivered_carrier_date,1783,1.793023
n_items,775,0.779357
price,775,0.779357
freight,775,0.779357
review_score,768,0.772317
order_approved_at,160,0.160899
payment,1,0.001006
installments,1,0.001006
payment_type,1,0.001006


In [16]:
df.duplicated().sum()

0

In [17]:
df["order_id"].duplicated().sum()

0

In [18]:
df["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [19]:
df["payment_type"].value_counts(dropna=False)

payment_type
credit_card    75387
boleto         19784
voucher         2739
debit_card      1527
not_defined        3
NaN                1
Name: count, dtype: int64

In [20]:
df["review_score"].value_counts(dropna=False).sort_index()

review_score
1.000000    11316
1.500000        8
2.000000     3125
2.500000       34
3.000000     8136
3.333333        1
3.500000       25
4.000000    19018
4.333333        1
4.500000       54
5.000000    56955
NaN           768
Name: count, dtype: int64

In [21]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

df[date_columns].head()

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


##  Initial Observations

### Dataset Overview
- The dataset contains **99,441 order-level records and 15 variables** covering order details, customer identifiers, payment information, delivery timestamps, and review scores.
- `order_id` and `customer_id` are unique in this prepared dataset, indicating that each row represents a distinct order record.

### Data Quality
- No completely duplicated rows were identified.
- Missing values are present in several columns, with the highest proportion found in `order_delivered_customer_date` (approximately 2.98%).
- Missing delivery-related dates may be associated with orders that were not successfully delivered and will require further investigation before treatment.
- `n_items`, `price`, and `freight` each have approximately 0.78% missing values, while `review_score` has approximately 0.77% missing values.

### Order and Payment Overview
- The majority of orders have a `delivered` status, accounting for approximately 97% of the dataset.
- `credit_card` is the most frequently recorded payment method, followed by `boleto`.
- Review scores range from 1 to 5, providing an opportunity to study customer satisfaction in relation to order and delivery characteristics.

### Data Preparation Requirements
- The date-related columns are currently stored as non-datetime values and will require appropriate datetime conversion during preprocessing.
- Missing values will be investigated based on their business context rather than being removed or filled blindly.
- Numerical variables such as `price`, `freight`, and `payment` show wide ranges and will require further distribution and outlier analysis.

### Next Steps
- Investigate the causes and patterns of missing values.
- Convert and validate date/time columns.
- Perform data cleaning and preprocessing.
- Create useful analytical features such as delivery duration and delivery delay.
- Continue with purchase, payment, delivery, and customer satisfaction analysis.